File: developUtils/urlDuplicateRemoval.js
import urlList from '../locales/json/zh-cnallUniversalLink.json' with { type: 'json' };

// 这个工具用来查找 allUniversalLink.json 里重复的 url

function findAndDeduplicateUrls(obj) {
  let urlValues = []; // 用来存储所有url的值
  let duplicates = new Set(); // 用来存储重复的url

  // 遍历对象的所有属性
  const traverse = (curr) => {
    for (let key in curr) {
      if (Object.prototype.hasOwnProperty.call(curr, key)) {
        // 如果key是url，收集url的值
        if (key === 'url' && typeof curr[key] === 'string') {
          if (urlValues.includes(curr[key])) {
            // 如果已经存在这个url值，标记为重复
            duplicates.add(curr[key]);
          } else {
            urlValues.push(curr[key]);
          }
        } else if (typeof curr[key] === 'object' && curr[key] !== null) {
          // 如果是嵌套对象，继续递归
          traverse(curr[key]);
        }
      }
    }
  };

  // 开始递归遍历
  traverse(obj);

  // 输出重复的url值
  duplicates.forEach((url) => {
    console.log(url);
  });
}

// 执行函数
findAndDeduplicateUrls(urlList);

File: developUtils/urlDuplicateRemoval.py
import json
import os
import sys

def find_and_deduplicate_urls(obj):
    url_values = []  # 存储所有URL值
    duplicates = set()  # 存储重复的URL

    # 递归遍历函数
    def traverse(curr):
        if isinstance(curr, dict):
            for key, value in curr.items():
                if key == 'url' and isinstance(value, str):
                    if value in url_values:
                        duplicates.add(value)
                    else:
                        url_values.append(value)
                else:
                    traverse(value)
        elif isinstance(curr, list):
            for item in curr:
                traverse(item)

    # 开始递归遍历
    traverse(obj)

    # 输出重复的URL
    for url in duplicates:
        print(f'{url}')

def main(top_directory='../locales'):
    if not os.path.isdir(top_directory):
        print(f"错误: 顶层目录 '{top_directory}' 不存在")
        sys.exit(1)
        
    subdirs = []
    for item in os.listdir(top_directory):
        item_path = os.path.join(top_directory, item)
        if os.path.isdir(item_path):
            subdirs.append((item, item_path))
        elif item.endswith('.js'):
            continue
        else:
            print(f"警告: 忽略非目录非js文件: {item}")
            
    if not subdirs:
        print("错误: 没有找到次顶层目录")
        sys.exit(1)

    print(f"找到 {len(subdirs)} 个次顶层目录: {[name for name, path in subdirs]}")
    
    for name, path in subdirs:
        with open(f'{path}/json/allUniversalLink.json', 'r', encoding='utf-8')as f:
            ref_json = json.load(f)
        print(f'检测{path}/json/allUniversalLink.json')
        find_and_deduplicate_urls(ref_json)

if __name__ == '__main__':
    if len(sys.argv) > 2:
        print("用法: python script.py <顶层目录 default=../locales> ")
        sys.exit(1)
    if len(sys.argv) >1:
        main(sys.argv[1])
    else:
        main()